src/App.tsx

```diff
@@ -1,10 +1,139 @@
+
+import {
+  addTodo,
+  deleteTodo,
+  getTodos,
+  updateTodoStatus,
+  USER_ID,
+} from './api/todos';
+
+import { TodoList } from './components/TodoList';
+import { Footer } from './components/Footer';
+import { ErrorNotification } from './components/ErrorNotification';
+import { Header } from './components/Header';
+
+import { Todo } from './types/Todo';
+
+export enum FilterTypes {
+  All = 'All',
+  Active = 'Active',
+  Completed = 'Completed',
+}
+
+export enum ErrorMessage {
+  none = '',
+  load = 'Unable to load todos',
+  emptyTitle = 'Title should not be empty',
+  add = 'Unable to add a todo',
+  delete = 'Unable to delete a todo',
+  update = 'Unable to update a todo',
+}
+
+function filteringTodos(arrayOfTodos: Todo[], filterType: FilterTypes) {
+  switch (filterType) {
+    case FilterTypes.Active:
+      return arrayOfTodos.filter(item => !item.completed);
+    case FilterTypes.Completed:
+      return arrayOfTodos.filter(item => item.completed);
+    case FilterTypes.All:
+      return arrayOfTodos;
+  }
+}
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [newTodoTitle, setNewTodoTitle] = useState<string>('');
+  const [filterType, setFilterType] = useState<FilterTypes>(FilterTypes.All);
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const [isLoading, setIsLoading] = useState(false);
+  const [errorMessage, setErrorMessage] = useState<ErrorMessage>(
+    ErrorMessage.none,
+  );
+
+  useEffect(() => {
+    setErrorMessage(ErrorMessage.none);
+    setIsLoading(true);
+
+    getTodos()
+      .then(setTodos)
+      .catch(() => setErrorMessage(ErrorMessage.load))
+      .finally(() => setIsLoading(false));
+  }, []);
+
+  const deleteSelectedTodo = (targetId: number) => {
+    setErrorMessage(ErrorMessage.none);
+    setIsLoading(true);
+    setTodos(currentTodos => currentTodos.filter(todo => todo.id !== targetId));
+
+    deleteTodo(targetId)
+      .catch(() => {
+        setTodos(todos);
+        setErrorMessage(ErrorMessage.delete);
+      })
+      .finally(() => setIsLoading(false));
+  };
+
+  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    setErrorMessage(ErrorMessage.none);
+
+    const pushingNewTodo: Omit<Todo, 'id'> = {
+      userId: USER_ID,
+      title: newTodoTitle,
+      completed: false,
+    };
+
+    if (newTodoTitle) {
+      setIsLoading(true);
+
+      return addTodo(pushingNewTodo)
+        .then(newTodo => setTodos(currentTodos => [...currentTodos, newTodo]))
+        .catch(() => setErrorMessage(ErrorMessage.add))
+        .finally(() => {
+          setIsLoading(false);
+          setNewTodoTitle('');
+        });
+    }
+
+    return setErrorMessage(ErrorMessage.emptyTitle);
+  };
+
+  const isTodosEmpty = todos.length === 0;
+
+  const isAllTodosCompleted = todos.every(todo => todo.completed);
+
+  const filteredTodos = filteringTodos(todos, filterType);
```

Comment:
Wrap your calculations in useMemo. This can optimize performance by memoizing the results, ensuring that the computations are only re-executed when their dependencies change.